### ISSUE TITLE
[ECS] Optional component for `Instance` used to control time, weather, world border, etc.

### ISSUE THREAD

**Issue by rj00a** [valence-rs/valence#210]
It would be convenient if `Instance` had some optional components for controlling state which is typically expected to be global to the instance. This includes the time of day, whether or not time is paused (`doDaylightCycle`), The world border for the instance, the weather (raining or thundering), etc.

My current thinking is that we can have the following components:
- `TimeOfDay`
- `Weather`
- `WorldBorder`

**Comment by rj00a** [comment]
> Might be more convenient to make these components smaller for the purpose of change detection.

Yes, each field that needs change detection should be its own component.

**Comment by dyc3** [comment]
I'd consider that to be vanilla functionality, which would be out of scope for valence. Do gamerules get conveyed to clients and affect them in any way? If not, gamerules are definitely out of scope.

**Comment by arseeeeN** [comment]
That makes sense, so the developers that use Valence would handle the updating of the time of day themselves? Including sending that information to the players of course.
Should the time of day be a field of the `InstanceInfo` struct or a component that could be spawned alongside the entity of the Instance component?

**Comment by dyc3** [comment]
Valence would be the one sending packets. As the issue says, the desired solution is a `TimeOfDay` component that goes on the entity that has the `Instance`. A system inside valence would then be able to handle sending those packets.

**Comment by MrlnHi** [comment]
> I'd consider that to be vanilla functionality, which would be out of scope for valence. Do gamerules get conveyed to clients and affect them in any way? If not, gamerules are definitely out of scope.

reducedDebugScreen or however it's called gets conveyed to clients (and actually makes a difference). Other than that I can't think of any tho

**Comment by tachibanayui** [comment]
Is anyone already working on time of day? I want to take a shot at implementing it.

**Comment by arseeeeN** [comment]
I initially wanted to work on it but couldn't find the time, so feel free to take over.

**Comment by 1bacon** [comment]
> > I'd consider that to be vanilla functionality, which would be out of scope for valence. Do gamerules get conveyed to clients and affect them in any way? If not, gamerules are definitely out of scope.
> 
> reducedDebugScreen or however it's called gets conveyed to clients (and actually makes a difference). Other than that I can't think of any tho

For completeness:
Only `reducedDebugInfo`, `doLimitedCrafting` and `doImmediateRespawn` are sent to the Client.

All others are checked by the server internally.
